### PR TITLE
Add permissions to kibana_system for TI package transforms to support IOC expiration

### LINKED ISSUE #94505: Need kibana_system privileges for Threat Intel (TI) package transforms to support IOC expiration
### Description

**Requirement:**
Threat Intelligence (TI) data providers generate data in the form of indicators of compromise (IOCs). The threat landscape is always evolving and therefore the IOCs need to update to reflect the current state or expired when the indicators are no longer relevant.

As a user of indictor match rules I only want to match against the latest IOCs only.
As a threat intel provider I want to be able to update or expire IOCs that my customers store.

**Support:**
- Using `latest` transform in Elasticsearch to maintain "latest" index that only contains latest IOCs and not the expired ones.

**Current Issue:**
- In current scenario, when installing the integration with the transform, following error is received: `Error installing ti_recordedfuture 1.6.0: security_exception: [security_exception] Reason: action [indices:admin/create] is unauthorized for service account [elastic/kibana] on indices [logs-ti_recordedfuture_latest.threat], this action is granted by the index privileges [create_index,manage,all]`. Reference: https://github.com/elastic/integrations/pull/5460
- It seems like the service account `elastic/kibana` which is automatically used for installing the integration/assets doesn't seem to contain necessary privileges required on destination index, which needs `index, read, create_index, delete` permissions. It also needs a `read` on source indices.
- The other permissions, namely `view_index_metadata` and `manage_transform` are already added to the `elastic/kibana` service account.

Need to[ setup required permissions onto ](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java#L682)`kibana_system` on all TI destination indices required to support IOC expiration using transforms.

### PR DESCRIPTION
**Pull request by kcreddy**
Closes https://github.com/elastic/elasticsearch/issues/94505 by adding required permissions to `kibana_system` to support IOC expiration of Threat Intel indices.

### CODE REVIEW

**Comment by ywangd** [comment]
Ping @elastic/kibana-security for awareness

**Review by kc13greiner — DISMISSED** [review]
~~LGTM~~ Discussing with KB Security team!

**Comment by kc13greiner** [comment]
> Ping @elastic/kibana-security for awareness

ACK! Thanks for the update!

**Comment by P1llus** [comment]
Just before we merge, @andrewkroh, would it be wise to have a pattern outside of logs-*? If we want to place all our TI transforms there in the future, and any future logs-* templates and global index patterns might complicate things if it covers the transform indices as well?

**Review by kc13greiner — CHANGES_REQUESTED** [review]
Would it be possible to change the name of the indices being used so that they have a `.` preceding? This would be more in line with other system indices. (like the indices listed at the bottom of this [comment](https://github.com/elastic/security-team/issues/6114#issuecomment-1476416254)

As is, this change may allow `kibana_system` to access user defined indices that follow a similar pattern.

Generally we would like to prevent the `kibana_system` user from getting too powerful.

cc: @azasypkin @legrego

**Comment by kcreddy** [comment]
Hey @andrewkroh,
As per @kc13greiner review above, I changed names of the indices with a preceding `.`. Could you please confirm if this also addresses @P1llus concern [above](https://github.com/elastic/elasticsearch/pull/94506#issuecomment-1489992516)? Or does it make sense to move into a different pattern altogether? If so, may I have a suggestion please? Thanks!

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java
Comment (andrewkroh) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:863]:
```suggestion
                    .indices(".logs-ti_*_latest.*")
```

I'm trying to be as specific as possible.

**Review by andrewkroh — COMMENTED** [review]
> Would it be possible to change the name of the indices being used so that they have a `.` preceding? This would be more in line with other system indices.

@kc13greiner It's not really a "system index" as I have seen the term defined in various Elasticsearch issues. It is going to contain data that users directly access. Users will utilize it through the Security [Indicator Match](https://www.elastic.co/guide/en/security/current/rules-ui-create.html#create-indicator-rule) Rules, the Security [Intelligence](https://www.elastic.co/guide/en/security/current/es-ui-overview.html#_intelligence) page, and possibly Discover. Therefore I think it would be better to not use a `.` to indicate this is a system or hidden.

I view this as a workaround until a proper mechanism exists for authorizing the transform setup based on the user installing the Fleet integration. That feature is described in https://github.com/elastic/kibana/issues/137278, but AFAIK is not started yet. I would hope that we could remove the privileges directly related to transform setup after that task is completed.

These other changes have added similar workarounds:

- https://github.com/elastic/elasticsearch/pull/84941
- https://github.com/elastic/elasticsearch/pull/94255

But if a dot is required and users will still be able to directly read the data through the aforementioned apps then a dot prefix can work.

**Comment by P1llus** [comment]
> > would it be wise to have a pattern outside of logs-*? If we want to place all our TI transforms there in the future, and any future logs-* templates and global index patterns might complicate things if it covers the transform indices as well?
> 
> @P1llus What specifically is the concern that could cause a complication? And by "to place all our TI transforms there" do you mean the destination index name for the installed transforms? What naming conventions are you thinking?
> 
> In https://github.com/elastic/integrations/pull/5582 the convention used is `logs-ti_<name>_latest.<data_stream>` for transform destination and the integration data streams are `logs-ti_<name>.<data_stream>-*`.
> 
> IIUC Fleet should create an index template for the transform's target index. This is described at https://github.com/elastic/package-spec/blob/29cc306237b541e7c32841548c09e25dcfd19333/spec/integration/elasticsearch/transform/spec.yml#L7-L10. That template should have a priority of 200 according to [this](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-templates.html). So as long as the transforms destination index names don't overlap with the integration data stream naming it shouldn't be an issue. I think the inclusion of `_latest` should work to prevent collisions.
> 
> I'm not opposed to a completely non-overlapping naming convention, but there is no prior art within Fleet that I'm aware of. 

@andrewkroh I was just concerned that any future ILM, index templates, dynamic templates or other settings that might be added to logs-* as a global pattern could affect the destination index for the transforms, outside of that I do not have any other concerns.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java
Comment (kcreddy) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java]: Updated as suggested

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java
Comment (kcreddy) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java]: Removed `update settings, put mapping, and rollover` as suggested

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java
Comment (kcreddy) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:881]: Added separate comments for ILM and transform permissions

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java
Comment (kcreddy) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java]: Updated as suggested

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java
Comment (kcreddy) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java]: Updated as suggested

**Comment by kc13greiner** [comment]
> Hey @kc13greiner
> 
> > @kc13greiner It's not really a "system index" as I have seen the term defined in various Elasticsearch issues. It is going to contain data that users directly access. Users will utilize it through the Security [Indicator Match](https://www.elastic.co/guide/en/security/current/rules-ui-create.html#create-indicator-rule) Rules, the Security [Intelligence](https://www.elastic.co/guide/en/security/current/es-ui-overview.html#_intelligence) page, and possibly Discover. Therefore I think it would be better to not use a . to indicate this is a system or hidden.
> 
> The current code has the `.` prefix added. Based on Andrew's suggestion above, could you give your confirmation that its required or not? Could you please review and let me know if thats the only change required?

@kcreddy @andrewkroh 

I have discussed with the security team and we are ok to approve without the `.` prefix on the indices. The [docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-templates.html) do warn users against using these `log-` patterns to avoid collisions.

Sorry for the confusion.

**Comment by kcreddy** [comment]
@kc13greiner,
Sorry for the delay. I removed the dot `.` prefix. Could you please approve/merge if everything looks good?

**Review by andrewkroh — APPROVED** [review]

**Review by kc13greiner — APPROVED** [review]
LGTM!

**Pull request merged**
